### ISSUE TITLE
DecompressionStream exception on dispose

### ISSUE THREAD

**Issue by alexanderzuban** [oleg-st/ZstdSharp#4]
DecompressionStream raises an exception on disposal:

{{{
...
            if (lastDecompressResult != 0)
                throw new EndOfStreamException("Premature end of stream");
...
}}}

Because the class has a finalizer which is calling Disposal this may lead to exception/crash during finalization, which is impossible to handle/control. It will be great to change it to avoid crashes during GC. 

{{{
System.IO.EndOfStreamException: Premature end of stream
DecompressionStream.Dispose (System.Boolean disposing)
DecompressionStream.Finalize ()
}}}

Also, it may have the sense to always release input stream buffer during disposal, to avoid possible memory leaks (I didn't dive deep what does lastDecompressResult means if this is a flag that we still have something to decompress, it is possible that the reader just stop to read from the stream for some reason, but disposal should release buffer in this case anyway)

**Comment by oleg-st** [comment]
Yes, nonzero `lastDecompressResult` means we are not at the end of the frame. It should be checked after `innerStreams`'s end, not in Dispose / Finalize. It's means premature end of `innerStream`.

**Comment by alexanderzuban** [comment]
but this will lead to mem leak in the general case(stream will never release pooled buffer) and app crash if stream usage is missing using(this is bad, but the app doesn't expect GC resource cleanup will crash)

**Comment by oleg-st** [comment]
Fixed in 0.5.2

**Comment by oleg-st** [comment]
@alexanderzuban 
Could you check it?

**Comment by alexanderzuban** [comment]
Seems to work fine now, thank you!
